text/2020-03-11-versioned-kv.md

```diff
@@ -0,0 +1,202 @@
+
+#### Data model and Consistency
+
+The data model is like `key:version -> value`. The version uses the global `ts`
+obtained from the PD TSO by default. Then strong consistency (single key level
+ linearizability) would be support. And the implementation is subject to
+ further change if we use the user TS feature of RocksDB.
+
+The maximum number of versions in the cluster can be configured via `MaxVerNum`.
+The default value is 1. Redundant versions are removed during version
+recycling (compaction or garbage collection).
+
+#### Data storage
+
+`user_key` is encoded according to Memory Comparable Encoding in the TiKV client
+sdk, and the encoding algorithm is padding in 8-byte units. This operation
+ ensures that ts is appended to the key and the written user_key (key: ts)
+can keep the same order as the key itself.
+
+The timestamp stored in user_key is its inversion result. The purpose is
+to arrange the newer data (that is, the data with larger timestamp) in front
+ of the older data (that is, data with smaller timestamp).
+This feature can be utilized in performance optimizing for the data
+scan process.
+
+#### Data distribution
+
+To facilitate the acquisition and efficient management of multiple versions of
+a key, the split point of the region will not use the timestamp of the key
+, so that multiple versions of the key can be guaranteed to be
+in the same region and the same CF. Considering the size of the region,
+MaxVerNum will be restrained.
+
+The process of data balancing, scaling, and migration are the same as that
+of RawKV. In the process of scanning all versions of a key, if the maximum
+ number of versions is exceeded, the redundant versions will be discarded.
+
+#### RPC Interface
+
+With reference to the proto interface, the following interfaces are defined as
+a new service other than tikv:
+
+- Get/BatchGet - Gets the values for latest version of the key by default
+- GetNum/BatchGetNum - Get the values for a specified number of versions of a
+  key. `version_num` must be less than or equal to `MaxVerNum`.
+- GetNumWithSV/BatchGetNumWithSV - This interface is like `GetNum/BatchGetNum`,
+  but add requirement about the key with versions ranging from the specified
+  start version to the current version.
+- GetWithTTL/BatchGetWithTTL - Gets the unexpired values within the TTL.
+- Scan - Scans the values of all keys of the latest version within a range
+- ScanNum - Scans the number of all keys within a range to get the
+  corresponding values of a specified number of versions
+- ScanNumWithSV - Scans all the keys within a range and gets the
+  corresponding values ranging from the specified start version to the current
+  version
+- Put/BatchPut - Increases the new version value of a key, where version is
+  the current timestamp by default
+- Delete/BatchDelete - Deletes all versions of a specified key
+  (`key:cur_ts -> delete`)
+- DeleteRange - Deletes all versions of key in a range using rocksdb DeleteRange
```

Comment:
Could we start with a more minimal API and then add some of these later once we have some experience with the API? One way to do that is to prioritize the commands, and we'd implement the p1 commands, then wait a bit, then the p2 commands, etc.

Some orthogonal questions:
* could the four `get` commands be combined by having a max num, min version, and TTL arguments? Likewise with Scan.
* I still think that the GetWithTTL command is not needed: e.g., if the current ts is 1000 and the client wants GetWithTTL with a TTL of 100, then the client could call GetNumWithSV with ts = 900 (and num is some large number).
* The description of `Put` is confusing to me. Do you mean "adds a new value for a key with a given timestamp"? (As mentioned elsewhere, I don't think TiKV should know the current time).

Reply:
Ok, I would modify about `get` and remove `GetWithTTL`. for `Put`, the timestamp is obtained from the PD TSO when the client sdk is doing the request .